From a19a97464a0292d38496b96bce993d5751578829 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Thu, 09 Mar 2023 08:53:31 +0200
Subject: [PATCH] [SW-127242] gaudi2: erom: fix boot failure in cold reset

The CCM memories gets cleared in cold reset hence
accessing ARC1 ICCM for copying boot info can cause
ecc exception causing ARC to get hung. This patch
adds support to initialize the ARC1 ICCM in cold
reset in erom recovery flow.

Also this patch takes care to avoid copying boot_info
from persistent to spsram during cold boot.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I7767199ba86c5952002c5b8bcec736aa8d06df62
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 2efd514..cf529b6 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -1037,6 +1037,14 @@
 
 void preboot_boot_info_init(void)
 {
+	u32 rst_src = hl_rst_src_get();
+
+	if (rst_src == HL_COLD_RST) {
+		/* Scrub ARC1 ICCM as it is used for boot info to avoid ecc errors */
+		init_arc_mem(1, ARC1_ICCM_FULL_BASE, ARC1_ICCM_SIZE);
+		return;
+	}
+
 	/* copy boot info from persistent memory to SPSRAM */
 	boot_info_copy_from_perst_mem();
 	/* boot info finalize does not merge old data hence copy and set it explicitly */
